### ISSUE TITLE
In disclosure nav menu example, when screen reader in reading mode, arrows should move through submenus

### ISSUE THREAD

**Issue by RichCaloggero** [w3c/aria-practices#1243]

**Comment by RichCaloggero** [comment]
https://www.w3.org/TR/wai-aria-practices-1.1/examples/disclosure/disclosure-navigation.html

- open a menu by focusing then pressing space or enter (for example, open the "about" menu)
- press downArrow when screen reader in document reading mode
   + should focus "overview"
   + actual result: focus moves to next sibling of "about": "Admissions"

**Comment by JAWS-test** [comment]
I suspect that this is an error of the screenreader. The example is correct because the submenu entries in the source code follow immediately after the respective menu entry.
I tested it with JAWS 2019 and Chrome. Sometimes I could use the arrow keys to read the submenu after it was displayed, sometimes I also experienced the behavior you described.

**Comment by JAWS-test** [comment]
In my opinion, however, the following sentence in the example is wrong because there is no content that automatically fades in on focus or hover:

> Implementing this Esc behavior is necessary to meet the WCAG 2.1 1.4.13: Content on Hover or Focus criterion.

I would also like to point out that the example does not work in IE 11 (ESC and arrow key operation). The reason seems to be that IE does not support `repeat` in JavaScript. The arrow icons are also not displayed correctly in IE 11.

**Comment by ZoeBijl** [comment]
@RichCaloggero which screen reader and browser did you use?

The example you pointed to seems to work fine in Chrome 78 with VoiceOver on macOS 10.14.6.

**Comment by JAWS-test** [comment]
> which screen reader and browser did you use?

The problem occurs for example with JAWS and Chrome.

**Comment by ZoeBijl** [comment]
> > which screen reader and browser did you use?
> 
> The problem occurs for example with JAWS and Chrome.

Thank you. But I’d like to know hear from @RichCaloggero what they used.

**Comment by RichCaloggero** [comment]
Windows and Apple (both mac and iOS) use two different ways of handling web content.

Apple has only a single mode (see below) and thus all operations occur on the live DOM.

Windows screen readers use two modes: reading (document or sometimes virtual) mode and live (or real) mode. In live mode the screen reader interacts directly with the browser, but can only interact with focusable elements, and only tab navigation is possible because it moves among focusable elements. In virtual mode, the screen reader interacts with it's own virtual DOM and heading navigation (as well as navigation by many other element types including arrow navigation which moves through everything similar to the way voiceOverKeys+rightArrow does) is supported.

I'll need to dig into the javascript here, but my best guess at the moment is that the arrow keys are being captured by a keydown handler somewhere which makes them unavailable to NVDA (and probably jaws as well), even in virtual mode. The VoiceOver key combination VO+rightArrow, however, is not handled by the app so it does what is expected.

-- Rich

**Comment by RichCaloggero** [comment]
## CSS?

I rewrote the js removing all keyboard handling, just retaining toggling behavior. When I apply the original CSS to the original html, the issue still occurs.

I narrowed it down to this rule:

```
.disclosure-nav button:focus {
  border-color: #005a9c;
  outline: 5px solid rgba(0, 90, 156, 0.75);
  /*position: relative;*/
}
```

However, being totally blind, I do not understand the CSS. Would some kind guru please explain what this is doing? Can the CSS berewritten to remove this rule?

**Comment by JAWS-test** [comment]
´position: relative;´ is used so that the focus highlighting with ´outline´ is not hidden by the next element (to the right of it).

Is the ´position: relative;´ the cause for the incorrect output by the screenreader?

**Comment by RichCaloggero** [comment]
Yes, the `position:relative` is what is effecting the screen reader (NVDA+Firefox). I will test with other combos when I get to work today and report back.

Is there another way of implementing this which does not use `position:relative`?

More later today...

**Comment by RichCaloggero** [comment]
## More extensive testing

- fails with firefox 70.0.1 and NVDA
- succeeds with all other combos I've tried
   + nvda with firefox-72.0.A1 (nightly)
   +  nvda with firefox-71.0.b9 (developer edition)
   + jaws with firefox-70.0.1
   + nvda with chrome

Seems like a bug in current firefox release.

**Comment by RichCaloggero** [comment]
## For completeness

When viewing the following code in Firefox-70.0.1 with latest NVDA (2019.2.1), the focusable span is announced, but the sublist is completely missing.

Oddly enough, if you look at the accessibility tree via Firefox devtools, the accessible name on the sublist `li` is present, which means the screen reader should see it.

Very odd indeed.

Again, however, this does seem to be fixed in firefox beta and alpha releases!

```
<!doctype html>
<html>
<head>
<meta charset="utf-8">
<title>test</title>
<style>
ul {list-style-type: none;}
ul .button:focus {position: relative;}
</style>
</head>
<body>

<nav>
<ul><li>
<span tabindex="0" class="button">A Button</span>
<ul><li>a sublist item</li></ul>
</li></ul>
</nav>

</body>
</html>
```

**Comment by JAWS-test** [comment]
I created the [Example Disclosure for Navigation menus](https://codepen.io/jaws-test/pen/wvvEmaB) without `position:relative` and the error does not occur in Firefox+NVDA. The focus display is not correct (even if `position:relative` is permanently assigned to the buttons and not only at `:focus`), but can certainly be displayed in other ways. It seems to me to be a problem of NVDA, so I think the ticket should rather be opened at NVDA.

**Comment by RichCaloggero** [comment]
Have you tested in beta and alpha versions of firefox? Those seem to work fine with NVDA, so I think it may be an issue with CSS processing in firefox which has been fixed in upcoming versions.

**Comment by JAWS-test** [comment]
> Have you tested in beta and alpha versions of firefox? 

I have only tested with my standard version of Firefox

> may be an issue with CSS processing in firefox which has been fixed

I hope so

**Comment by mcking65** [comment]
I notice that a different problem occurs with JAWS in Firefox. after you collapse a menu, the list of links does not go away until you press insert+escape. At any rate, I agree this is a firefox bug based on all the analysis here, and it is good that they already have a fix on the way out the door.

Rich, thank you for the great analysis. Both amazing and kind of scarey how fragile accessibility support is. I can't wait until we have solid automated testing of all of this.

Closing the issue.
